reword to include other objects and transitive mutations (#1)
stephanietuerk/svelte: site/content/tutorial/02-reactivity/04-updating-arrays-and-objects/text.md

title: Updating arrays and objects
---

Svelte's reactivity is triggered by assignments. Methods that mutate arrays/objects internally will not trigger updates by themselves.

In this example, clicking the "Add a number" button calls the `addNumber` function, which appends a number to the array but doesn't trigger the recalculation of `sum` and does not update the rendered page.

> One way to fix that would be to assign `numbers` to itself to tell the compiler it has changed:

Reviewer: ```suggestion
One way to fix that is to assign `numbers` to itself to tell the compiler it has changed:
```

}
```

You can also get a similar result using the ES6 spread syntax:

```js
function addNumber() {
	numbers = [...numbers, numbers.length + 1];
}
```

> The same rule applies for other array methods such as `pop`, `shift`, `splice` and for other objects such as `Map.set`.

Reviewer: ```suggestion
The same rule applies to array methods such as `pop`, `shift`, `splice` and to objects methods such as `Map.set`.
```

}
```

However, indirect assignments to references such as this...

```js
const foo = obj.foo;
foo.bar = 'baz';
```

or 

```js
function quox(thing) {
	thing.foo.bar = 'baz';
}
quox(obj);
```

...won't trigger reactivity on `obj.foo.bar`, unless you follow it up with `obj = obj`.

> A simple rule of thumb: the updated variable must directly appear on the left hand side of the assignment.

Reviewer: ```suggestion
A simple rule of thumb: the name of the updated variable must directly appear on the left hand side of the assignment.
```

Author: I removed this to prevent ambiguity in cases like this:
```js
function quox(obj) {
	obj.foo.bar = 'baz'; // won't trigger updates on obj
}
quox(obj);
```

Are you certain this should be reverted?

Reviewer: oh hmm..
